File: src/main/java/tech/sirwellington/alchemy/http/mock/MockAlchemyHttp.java
/*
 * Copyright © 2018. Sir Wellington.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 *
 * You may obtain a copy of the License at
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package tech.sirwellington.alchemy.http.mock;

import java.util.*;
import java.util.concurrent.Callable;
import java.util.concurrent.ConcurrentHashMap;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import sir.wellington.alchemy.collections.lists.Lists;
import tech.sirwellington.alchemy.annotations.access.Internal;
import tech.sirwellington.alchemy.annotations.designs.StepMachineDesign;
import tech.sirwellington.alchemy.arguments.AlchemyAssertion;
import tech.sirwellington.alchemy.arguments.assertions.Assertions;
import tech.sirwellington.alchemy.http.*;
import tech.sirwellington.alchemy.http.exceptions.AlchemyHttpException;
import tech.sirwellington.alchemy.test.Checks;

import static java.lang.String.format;
import static junit.framework.Assert.fail;
import static tech.sirwellington.alchemy.annotations.designs.StepMachineDesign.Role.MACHINE;
import static tech.sirwellington.alchemy.arguments.Arguments.*;
import static tech.sirwellington.alchemy.arguments.assertions.Assertions.instanceOf;
import static tech.sirwellington.alchemy.arguments.assertions.Assertions.notNull;
import static tech.sirwellington.alchemy.http.mock.MockRequest.ANY_BODY;
import static tech.sirwellington.alchemy.http.mock.MockRequest.NO_BODY;

/**
 *
 * @author SirWellington
 */
@Internal
@StepMachineDesign(role = MACHINE)
class MockAlchemyHttp implements AlchemyHttp
{

    private final static Logger LOG = LoggerFactory.getLogger(MockAlchemyHttp.class);

    private final Map<MockRequest, Callable<?>> expectedActions = new ConcurrentHashMap<>();

    private final List<MockRequest> requestsMade = Lists.create();

    MockAlchemyHttp(Map<MockRequest, Callable<?>> expectedActions)
    {
        Checks.Internal.checkNotNull(expectedActions);

        this.expectedActions.putAll(expectedActions);
    }

    @Override
    public AlchemyHttp usingDefaultHeader(String key, String value)
    {
        return this;
    }

    @Override
    public Map<String, String> getDefaultHeaders()
    {
        return Collections.emptyMap();
    }

    @Override
    public AlchemyRequestSteps.Step1 go()
    {
        return new MockSteps.MockStep1(this);
    }

    @Internal
    HttpResponse getResponseFor(MockRequest request) throws AlchemyHttpException
    {
        checkThat(request)
            .is(Assertions.<MockRequest>nonNullReference())
            .is(expectedRequest());

        requestsMade.add(request);

        Callable<?> action = findMatchingActionFor(request);

        Object response;
        try
        {
            response = action.call();
        }
        catch(AlchemyHttpException ex)
        {
            throw ex;
        }
        catch (Exception ex)
        {
            throw new AlchemyHttpException(ex);
        }

        checkThat(response)
            .usingMessage(format("Response Type Wanted: %s but actual: null", HttpResponse.class))
            .is(notNull())
            .usingMessage(format("Response Type Wanted: %s but actual: %s", HttpResponse.class, response.getClass()))
            .is(instanceOf(HttpResponse.class));

        return (HttpResponse) response;
    }

    @Internal
    <T> T getResponseFor(MockRequest request, Class<T> expectedClass) throws AlchemyHttpException
    {
        checkThat(request, expectedClass)
            .are(notNull());

        requestsMade.add(request);

        checkThat(request)
            .usingMessage("Unexpected Request: " + request)
            .is(expectedRequest());

        Callable<?> operation = findMatchingActionFor(request);

        Object responseObject;
        try
        {
            responseObject = operation.call();
        }
        catch(AlchemyHttpException ex)
        {
            throw ex;
        }
        catch (Exception ex)
        {
            throw new AlchemyHttpException(ex);
        }

        if (responseObject == null)
        {
            return (T) responseObject;
        }

        checkThat(responseObject)
            .usingMessage(format("Response Type Wanted: %s but actual: %s", responseObject.getClass(), expectedClass))
            .is(instanceOf(expectedClass));

        return (T) responseObject;

    }

    @Internal
    void verifyAllRequestsMade()
    {
        expected:
        for (MockRequest expectedRequest : expectedActions.keySet())
        {
            made:
            for (MockRequest requestMade : requestsMade)
            {
                if (requestsMatch(expectedRequest, requestMade))
                {
                    continue expected;
                }
            }

            //Reaching here means no match was found
            fail("Request never made: " + expectedRequest);
        }
    }

    private AlchemyAssertion<MockRequest> expectedRequest()
    {
        return request ->
        {
            checkThat(request)
                .is(Assertions.nonNullReference());

            Callable<?> action = findMatchingActionFor(request);

            Checks.Internal.checkNotNull(action, "");
        };
    }

    private Callable<?> findMatchingActionFor(MockRequest request)
    {
        Callable<?> foundInMap = expectedActions.get(request);

        if (foundInMap != null)
        {
            return foundInMap;
        }

        for (MockRequest element : expectedActions.keySet())
        {
            if (requestsMatch(element, request))
            {
                return expectedActions.get(element);
            }
        }

        return null;
    }

    private boolean requestsMatch(MockRequest expected, MockRequest actual)
    {
        boolean matchEverythingBesidesTheBody = matchEverythingBesidesTheBody(expected, actual);

        if (!matchEverythingBesidesTheBody)
        {
            return false;
        }

        if (expected.body == ANY_BODY)
        {
            return true;
        }

        if (expected.body == NO_BODY)
        {
            return actual.body == null || actual.body == NO_BODY;
        }

        /*
         * The bodies will be both null, or both set to NO_BODY. == is intentionally used to compare instances.
         */
        return expected.body == actual.body;
    }

    private boolean matchEverythingBesidesTheBody(MockRequest first, MockRequest second)
    {
        return Objects.equals(first.method, second.method) &&
               Objects.equals(first.url, second.url) &&
               Objects.equals(first.queryParams, second.queryParams);
    }

}

File: src/main/java/tech/sirwellington/alchemy/http/mock/MockAlchemyHttp.kt
/*
 * Copyright © 2018. Sir Wellington.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 *
 * You may obtain a copy of the License at
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package tech.sirwellington.alchemy.http.mock

import junit.framework.Assert.fail
import sir.wellington.alchemy.collections.lists.Lists
import sir.wellington.alchemy.collections.maps.Maps
import tech.sirwellington.alchemy.annotations.access.Internal
import tech.sirwellington.alchemy.annotations.arguments.Required
import tech.sirwellington.alchemy.annotations.designs.StepMachineDesign
import tech.sirwellington.alchemy.annotations.designs.StepMachineDesign.Role.MACHINE
import tech.sirwellington.alchemy.arguments.AlchemyAssertion
import tech.sirwellington.alchemy.arguments.Arguments.checkThat
import tech.sirwellington.alchemy.arguments.assertions.instanceOf
import tech.sirwellington.alchemy.arguments.assertions.nonNullReference
import tech.sirwellington.alchemy.arguments.assertions.notNull
import tech.sirwellington.alchemy.http.AlchemyHttp
import tech.sirwellington.alchemy.http.AlchemyRequestSteps
import tech.sirwellington.alchemy.http.HttpResponse
import tech.sirwellington.alchemy.http.exceptions.AlchemyHttpException
import tech.sirwellington.alchemy.http.mock.MockRequest.Companion.ANY_BODY
import tech.sirwellington.alchemy.http.mock.MockRequest.Companion.NO_BODY
import java.lang.String.format
import java.util.concurrent.Callable

/**
 *
 * @author SirWellington
 */
@Internal
@StepMachineDesign(role = MACHINE)
internal open class MockAlchemyHttp(expectedActions: Map<MockRequest, Callable<*>>) : AlchemyHttp
{

    private val expectedActions: MutableMap<MockRequest, Callable<*>> = Maps.createSynchronized()

    private val requestsMade = Lists.create<MockRequest>()

    override val defaultHeaders: Map<String, String>
        @Required
        get() = emptyMap()

    init
    {
        checkThat(expectedActions).isA(nonNullReference())
        this.expectedActions.putAll(expectedActions)
    }

    @Required
    override fun usingDefaultHeader(key: String, value: String): AlchemyHttp
    {
        return this
    }

    override fun go(): AlchemyRequestSteps.Step1
    {
        return MockSteps.MockStep1(this)
    }

    @Internal
    @Throws(AlchemyHttpException::class)
    open fun getResponseFor(request: MockRequest): HttpResponse
    {
        checkThat(request).isA(expectedRequest())

        requestsMade.add(request)

        val action = findMatchingActionFor(request)!!

        val response: Any
        try
        {
            response = action.call()
        }
        catch (ex: AlchemyHttpException)
        {
            throw ex
        }
        catch (ex: Exception)
        {
            throw AlchemyHttpException(ex)
        }

        checkThat(response)
                .usingMessage(format("Response Type Wanted: %s but actual: null", HttpResponse::class.java))
                .isA(notNull())
                .usingMessage(format("Response Type Wanted: %s but actual: %s", HttpResponse::class.java, response.javaClass))
                .isA(instanceOf(HttpResponse::class.java))

        return response as HttpResponse
    }

    @Internal
    @Throws(AlchemyHttpException::class)
    open fun <T> getResponseFor(request: MockRequest, expectedClass: Class<T>): T
    {
        checkThat(request, expectedClass)
                .are(notNull())

        requestsMade.add(request)

        checkThat(request)
                .usingMessage("Unexpected Request: " + request)
                .isA(expectedRequest())

        val operation = findMatchingActionFor(request)

        val responseObject: Any
        try
        {
            responseObject = operation!!.call()
        }
        catch (ex: AlchemyHttpException)
        {
            throw ex
        }
        catch (ex: Exception)
        {
            throw AlchemyHttpException(ex)
        }


        checkThat(responseObject)
                .usingMessage(format("Response Type Wanted: %s but actual: %s", responseObject.javaClass, expectedClass))
                .isA(instanceOf(expectedClass))

        return responseObject as T

    }

    @Internal
    open fun verifyAllRequestsMade()
    {
        expected@ for (expectedRequest in expectedActions.keys)
        {
            made@ for (requestMade in requestsMade)
            {
                if (requestsMatch(expectedRequest, requestMade))
                {
                    continue@expected
                }
            }

            //Reaching here means no match was found
            fail("Request never made: " + expectedRequest)
        }
    }

    private fun expectedRequest(): AlchemyAssertion<MockRequest>
    {
        return AlchemyAssertion { request ->

            checkThat(request).isA(nonNullReference())

            val action = findMatchingActionFor(request)
            checkThat(action).isA(nonNullReference())
        }
    }

    private fun findMatchingActionFor(request: MockRequest): Callable<*>?
    {
        val foundInMap = expectedActions[request]

        if (foundInMap != null)
        {
            return foundInMap
        }

        for (element in expectedActions.keys)
        {
            if (requestsMatch(element, request))
            {
                return expectedActions[element]
            }
        }

        return null
    }

    private fun requestsMatch(expected: MockRequest, actual: MockRequest): Boolean
    {
        val matchEverythingBesidesTheBody = matchEverythingBesidesTheBody(expected, actual)

        if (!matchEverythingBesidesTheBody)
        {
            return false
        }

        if (expected.body === ANY_BODY)
        {
            return true
        }

        return if (expected.body === NO_BODY)
        {
            actual.body == null || actual.body === NO_BODY
        }
        else
        {
            expected.body === actual.body
        }

        /*
         * The bodies will be both null, or both set to NO_BODY. == is intentionally used to compare instances.
         */
    }

    private fun matchEverythingBesidesTheBody(first: MockRequest, second: MockRequest): Boolean
    {
        return first.method == second.method &&
               first.url == second.url &&
               first.queryParams == second.queryParams
    }


}
